### ISSUE TITLE
Clipboard: allow to write to buffer from clipboard.write() method

### ISSUE THREAD

**Issue by bpasero** [electron/electron#11838]
* Electron version: 1.7.x
* Operating system: macOS

### Expected behavior

The method `clipboard.write()` accepts a data object but does not allow to write to a custom format with a Buffer. 

It would be great if this method could be changed to something like: 

```
clipboard.write({
  text: "some text",
  myFormat: new Buffer("some format")
});
```

My understanding is that the advantage of `clipboard.write()` is that I can set multiple clipboard contents at once so that the one format does not overwrite any other. Today, when I want to use `clipboard.writeBuffer()`, I cannot use it together with the other formats.

**Comment by danelkhen** [comment]
I'm having the same issue, relevant to Windows as well (I only see the platform/macOS label mentioned)

**Comment by itsananderson** [comment]
We looked into adding this API as part of the Electron Contributor Training. It appears that Chromium doesn't allow you to set "portable" and "platform" formats at the same time: https://source.chromium.org/chromium/chromium/src/+/master:ui/base/clipboard/scoped_clipboard_writer.cc;l=22?q=~ScopedClipboardWriter

We decided to try creating a `writeBuffers` API instead, which would look something like this:

```javascript
clipboard.writeBuffers({
    "public.utf8-plain-text": new Buffer("plain text"),
    "public.html": new Buffer("<b>plain text</b>"),
    "custom.format": new Buffer("some custom format")
});
```

This API wouldn't allow you to mix the "text", "html" etc. fields with your custom fields, so your code would need to know the actual platform formats for those generic types (as illustrated in the example above).

Still working on cleaning up my hacky first pass at the API, but I thought I'd post about the proposed API in case there's any feedback or suggestions.

**Comment by pdesantis** [comment]
@itsananderson that API sounds excellent! It sounds like it's still in development, but is your code pushed somewhere in a state you'd be comfortable sharing?

**Comment by muelphil** [comment]
Okay for anyone that searches for a **temporary** solution for **Windows-only** or a starting point the following is my native cpp code for implementation. Note that I am not a Cpp dev and this is most likely BAD CODE. Using the exported function in any other way than intended will just completly freeze your app.

index.cpp:
```c++
#include "napi-thread-safe-callback.hpp" // this also includes windows functions from winuser.h
// other includes

class Clipboard {
public:
	Clipboard(){
		OpenClipboard(0);
	}

	~Clipboard(){
		CloseClipboard();
	}
};

Napi::Boolean napiWriteBuffers(const Napi::CallbackInfo& info){
	Napi::Env env = info.Env();
	Napi::Value val = info[0];
	if(!val.IsObject()){
		return Napi::Boolean::New(env, false);
	}
	Napi::Object buffers = val.As<Napi::Object>();
	try{
		Clipboard* clipboard = new Clipboard();
		EmptyClipboard();
		Napi::Array properties = buffers.GetPropertyNames();
			for(uint32_t index = 0; index < properties.Length(); index++){
				Napi::Value property = properties[index];
				uint32_t propertyAsNumber = property.ToNumber().As<Napi::Number>().Uint32Value();
				Napi::Value bufferValue = buffers.Get(property);
				std::cout << "IsBuffer() = " << bufferValue.IsBuffer() << std::endl;
				Napi::Buffer<char> buffer = bufferValue.As<Napi::Buffer<char>>();
				auto bufferData = buffer.Data();
				SIZE_T size = sizeof(bufferData);
				UINT flags = GMEM_MOVEABLE;
				HGLOBAL handle = GlobalAlloc(flags, size);
				LPVOID data_ptr = GlobalLock(handle);
				memmove(data_ptr, bufferData, size);
				SetClipboardData(propertyAsNumber, handle);
				GlobalUnlock(handle);
			}
			return Napi::Boolean::New(env, true);
	} catch(const std::exception& e) {
		Napi::Error::New(env, e.what()).ThrowAsJavaScriptException();
		return Napi::Boolean::New(env, false);
	}
}

Napi::Object Init(Napi::Env env, Napi::Object exports) {
	exports.Set(
		Napi::String::New(env, "writeBuffers"),
		Napi::Function::New(env, napiWriteBuffers)
	);
	return exports;
}
```

and use like so:
```javascript
const mynativemodule= require( ... );
let result = mynativemodule.writeBuffers({
	15: Buffer.from('Hello', 'utf-8'), 
	49159: Buffer.from('World', 'utf-8')
});
console.log('result=', result);
```
Note that this solution only allows  1 argument of type `{[key:number]:Buffer}`. Lookup [Windows Clipboard Formats](https://docs.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats) for the numbers corresponding to the formats.

**Comment by zcbenz** [comment]
Does Chromium's new Asynchronous Clipboard API solve this issue?
https://w3c.github.io/clipboard-apis/#async-clipboard-api

**Comment by muelphil** [comment]
> Does Chromium's new Asynchronous Clipboard API solve this issue?
> https://w3c.github.io/clipboard-apis/#async-clipboard-api

Using this to read or write:
```javascript
// read
navigator.clipboard.readText()
// write
navigator.clipboard.write([new ClipboardItem({'CF_HDROP': new Blob(['Hello World'])})])
```

while the DOM/Window is not focused will in both cases throw
```error
Uncaught (in promise) DOMException: Document is not focused.
```

Therefore it only solves the issue for those reading from and writing to the clipboard while the application is active. For me the issue remains.

**Comment by issacgerges** [comment]
I've since moved companies but I believe this would have solved our issue assuming we could also enable [raw-clipboard](https://github.com/WICG/raw-clipboard-access/blob/main/explainer.md). I believe all non-standard types are still blocked by default [here](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/clipboard/clipboard_writer.cc;drc=40a8d5fea781057ff74a699247bc699f71ca4d5d;l=325)

**Comment by zcbenz** [comment]
@muelphil Would it solve the problem for you if we add an API to enable accessing `navigator.clipboard` when window is not focused?

**Comment by muelphil** [comment]
@zcbenz The main reason me and others are interested in a writeBuffers-function is to use it for formats other than text and html. In my case that would be CF_HDROP for copying windows files by their file location.

If I call the `navigator.clipboard.write` function within a `setTimeout` and manually refocus the document I sadly get

```
> setTimeout(() => navigator.clipboard.write([new ClipboardItem({'CF_HDROP': new Blob(['Hello World'])})]), 2000)
< Uncaught (in promise) DOMException: Type CF_HDROP not supported on write.
```

As ClipboardItem only accepts Mime-Types, this will not work for setting clipboard types other than those already supported by other methods of the electron clipboard module (text, html, rtf). Specifically platform dependent clipboard properties (multiple at once) cannot be set currently. Individually it is possibly using [clipboard.writeBuffer](https://www.electronjs.org/docs/api/clipboard#clipboardwritebufferformat-buffer-type-experimental), but calling this multiple times in a row will each time wipe the properties set before.

**Comment by issacgerges** [comment]
@muelphil. I might be misunderstanding you but see my comment above.

I think you need both what @zcbenz is proposing and `raw-clipboard` to be enabled.  You can try this locally by temporarily setting the flag in `chrome://flags. See https://github.com/WICG/raw-clipboard-access/blob/main/explainer.md#raw-clipboard-access-write for the API

**Comment by muelphil** [comment]
@issacgerges
Deeply sorry for the misunderstanding. I tried the proposed method like so:

```
setTimeout(() => {
navigator.clipboard.write([
  new ClipboardItem({'CF_HDROP': new Blob(['Hello'],{ type: 'CF_HDROP' })
  }, {raw:true}),
  new ClipboardItem({'FileName': new Blob(['World'],{ type: 'FileName' })
  }, {raw:true})
])}, 2000)
```
and I get
```
Uncaught (in promise) DOMException: Support for multiple ClipboardItems is not implemented.
```
It does work for a single ClipboardItem (namely CF_HDROP), however (additionally to not supporting multiple writes at once) it ignores the [Standard Clipboard Formats](https://docs.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats). CF_HDROP should be written to id 15, using [ClipView](http://www.peterbuettner.de/develop/tools/clipview/) I can observe that this it is registered as a new ClipboardFormat and has id 49702.

This is a minor inconvenience, the 'Support for multiple ClipboardItems is not implemented' is a showstopper.

**Comment by muelphil** [comment]
@issacgerges  Thanks for the heads up 👍 Dumb mistake on my side, although it is very misleading that this function takes an array if it is expecting a single argument.

This only leaves the issue with the wrong id (using the code you provided still registers a new clipboard format for `CF_HDROP` with a new id instead of id 15) and the `Document is not focused.`-error.

**Comment by zcbenz** [comment]
@muelphil Can you report the wrong id issue to Chromium? https://bugs.chromium.org/p/chromium/issues/detail?id=897289

Providing JavaScript APIs for writing native clipboard data is a very difficult project, since there is already a group of developers working on raw clipboard APIs, the fastest way to get the APIs is to help them, instead of working from scratch in Electron.

I'll add an API in Electron to solve the "Document is not focused" problem.

**Comment by muelphil** [comment]
I reported the issue some time ago, sadly it isn't resolved yet. ([link to the issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1204988#c1))

Out of interest, does this mean the electron clipboard API will be deprecated in favour for native chromium clipboard API?
For me atleast this would mean alot of refactoring, as the electron api takes Node `Buffer`s, while the chromium api uses `Blob`s. Using Node `Buffer`s in electron apps also seems more intuitive to me.

**Comment by muelphil** [comment]
The Issue was now assigned "WontFix (Closed)":
> Per https://crbug.com/1217643, raw clipboard is actually being deprecated+removed now, so I guess this bug is no longer applicable. Sorry it never did end up getting fixed properly.

So native Chromium solution isn't applicable to this issue anymore.

**Comment by zcbenz** [comment]
Thanks for letting us know, I'll start working on this once https://github.com/electron/electron/pull/26693 gets merged.

**Comment by henrit** [comment]
Another fallout of chromium removing support for its Raw Clipboard async API is that electron's `clipboard.writeBuffer()` calls are no longer working as intended when linked against a recent chromium (the clipboard ends up [getting "pickled"](https://github.com/w3c/editing/blob/gh-pages/docs/clipboard-pickling/explainer.md#pickling-for-async-clipboard-api) with some custom `application/web;type="custom/formatX"` and `application/web;type="custom/formatmap"` format names instead of the intended native format. )

Given this, the underlying implementation of Electron's `clipboard.writeBuffer` probably needs to be rewritten to no longer depend on `ScopedClipboardWriter`, but target `ui::Clipboard` directly instead. (And any hypothetical `clipboard.writeBuffers` method would need to use that too.)

**Comment by ArnonAsquira** [comment]
I understand that you can't write multiple custom formats to the clipboard using clipboard.writeBuffer(),
But when I try to use it write a "Google slides" format data, nothing is written to the clipboard:
("application/x-vnd.google-docs-drawings-page+wrapped" = Google slides format)
const slideBuffer = Buffer.from('slide format data');
clipboard.writeBuffer("application/x-vnd.google-docs-drawings-page+wrapped", slideBuffer);
and consequently, nothing is being pasted when
webcontent.paste() is called.

That being said, I can read the buffer from the clipboard using clipboard.readBuffer("application/x-vnd.google-docs-drawings-page+wrapped"),
but clipboard.availbaleFormtas() returns an empty array.

Are some formats not supported at all?

**Comment by YongzeYao** [comment]
any update? the `writeBuffers` seems very cool.

**Comment by MarshallOfSound** [comment]
`clipboard.write` takes `ClipboardItem`s with custom formats as of #51707 (44.x) so this is possible now.
